src/SignalR/server/Core/src/Internal/ChannelExtensions.cs

```diff
@@ -0,0 +1,36 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.SignalR.Internal;
+
+internal static class ChannelExtensions
+{
+    public static ValueTask RunAsync<TState>(this ChannelBasedSemaphore channelSemaphore, Func<TState, Task> callback, TState state)
+    {
+        if (channelSemaphore.AttemptWait())
+        {
+            _ = RunTask(callback, channelSemaphore, state);
```

Comment:
We should be delaying shutdown if the callbacks are still running. Is that happening? I have a feeling that `ChannelBasedSemaphore` will need to implement `IAsyncDisposable` or have some way to wait for all capacity to return before moving on.